File: backend/pkg/metrics/http/metrics.go
package http

import (
	"github.com/prometheus/client_golang/prometheus"
	"openreplay/backend/pkg/metrics/common"
	"strconv"
)

var httpRequestSize = prometheus.NewHistogramVec(
	prometheus.HistogramOpts{
		Namespace: "http",
		Name:      "request_size_bytes",
		Help:      "A histogram displaying the size of each HTTP request in bytes.",
		Buckets:   common.DefaultSizeBuckets,
	},
	[]string{"url", "response_code"},
)

func RecordRequestSize(size float64, url string, code int) {
	httpRequestSize.WithLabelValues(url, strconv.Itoa(code)).Observe(size)
}

var httpRequestDuration = prometheus.NewHistogramVec(
	prometheus.HistogramOpts{
		Namespace: "http",
		Name:      "request_duration_seconds",
		Help:      "A histogram displaying the duration of each HTTP request in seconds.",
		Buckets:   common.DefaultDurationBuckets,
	},
	[]string{"url", "response_code"},
)

func RecordRequestDuration(durMillis float64, url string, code int) {
	httpRequestDuration.WithLabelValues(url, strconv.Itoa(code)).Observe(durMillis / 1000.0)
}

var httpTotalRequests = prometheus.NewCounter(
	prometheus.CounterOpts{
		Namespace: "http",
		Name:      "requests_total",
		Help:      "A counter displaying the number all HTTP requests.",
	},
)

func IncreaseTotalRequests() {
	httpTotalRequests.Inc()
}

func List() []prometheus.Collector {
	return []prometheus.Collector{
		httpRequestSize,
		httpRequestDuration,
		httpTotalRequests,
	}
}

File: assist/utils/metrics.js
const client = require('prom-client')

// Create a Registry which registers the metrics
const register = new client.Registry()
register.setDefaultLabels({
    app: 'assist'
})

// Enable the collection of default metrics
client.collectDefaultMetrics({ register })

// http metrics

const httpRequestDuration = new client.Histogram({
    name: 'request_duration_seconds',
    help: 'A histogram displaying the duration of each HTTP request in seconds.',
    labelNames: ['method', 'route', 'code'],
    buckets: [.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 25, 50, 100, 250, 500, 1000],
});

const RecordRequestDuration = function(method, route, code, duration) {
    httpRequestDuration.observe({ method: method, route: route, code: code }, duration);
}

const httpTotalRequests = new client.Counter({
    name: 'requests_total',
    help: 'A counter displaying the number of all HTTP requests.',
});

const IncreaseTotalRequests = function () {
    httpTotalRequests.inc();
}

// websocket metrics

const websocketTotalConnections = new client.Counter({
    name: 'ws_connections_total',
    help: 'A counter displaying the number of all ws connections',
    labelNames: ['type'], // tab, agent
});

const IncreaseTotalWSConnections = function (type) {
    websocketTotalConnections.inc({type: type});
}

const websocketTotalRooms = new client.Counter({
    name: 'ws_rooms_total',
    help: 'A counter displaying the number of all rooms',
});

const IncreaseTotalRooms = function () {
    websocketTotalRooms.inc();
}


register.registerMetric(httpRequestDuration);
register.registerMetric(httpTotalRequests);
register.registerMetric(websocketTotalConnections);
register.registerMetric(websocketTotalRooms);

module.exports = {
    register,
    RecordRequestDuration,
    IncreaseTotalRequests,
    IncreaseTotalWSConnections,
    IncreaseTotalRooms,
}